### PR TITLE
Rule S1200: Classes should not be coupled to too many other classes

### LINKED ISSUE #515: Rule S1200: Classes should not be coupled to too many other classes (Single Responsibility Principle)
[RSPEC-1200](https://jira.sonarsource.com/browse/RSPEC-1200)

According to the Single Responsibility Principle, introduced by Robert C. Martin in his book "Principles of Object Oriented Design", a class should have only one responsibility:

Contributes to [MMF-963](https://jira.sonarsource.com/browse/MMF-963).

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #515

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+            return identifierName != null
+                ? new[] { model.GetSymbolInfo(identifierName).Symbol as ITypeSymbol }
+                : new[] { (ITypeSymbol)null };
+        }
+
+        private static int FilterAndCountCoupling(IEnumerable<ITypeSymbol> types)
+        {
+            return types.Distinct()
+                .WhereNotNull()
+                .Where(type => type.TypeKind != TypeKind.Enum &&
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: MAJOR Drop 'Where' and move the condition into the 'Count'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2971)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: MAJOR Drop 'Where' and move the condition into the 'Count'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2971)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+            KnownType.System_UInt32,
+            KnownType.System_Int64,
+            KnownType.System_UInt64,
+            KnownType.System_IntPtr,
+            KnownType.System_UIntPtr,
+            KnownType.System_Char,
+            KnownType.System_Single,
+            KnownType.System_Double,
+            KnownType.System_String,
+            KnownType.System_Object
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: Does FxCop count other special types, such as Func and Action? What about Task?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        }
+
+        private static IEnumerable<ITypeSymbol> CollectCoupledClasses(BaseMethodDeclarationSyntax method,
+            SemanticModel model)
+        {
+            return method.DescendantNodes()
+                .SelectMany(node => ExtractTypeSymbolsFromSyntaxNode(node, model));
+        }
+
+        private static IEnumerable<ITypeSymbol> ExtractTypeSymbolsFromSyntaxNode(SyntaxNode node, SemanticModel model)
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: perhaps if you use `yield return` in this method the code will look better: less created arrays, no need to return null in some cases (e.g. when parameter.Type == null), etc.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs (hunk 2 above)
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: Indeed, perhaps worth excluding all special types

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        };
+
+        protected override void Initialize(ParameterLoadingAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var classDeclaration = (ClassDeclarationSyntax)c.Node;
+
+                    if (c.IsTest() ||
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: Rspec doesn't say test are excluded, please update.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        }
+
+        private static IEnumerable<ITypeSymbol> ExtractTypeSymbolsFromSyntaxNode(SyntaxNode node, SemanticModel model)
+        {
+            var parameter = node as ParameterSyntax;
+            if (parameter != null)
+            {
+                return parameter.Type != null
+                        ? new[] { model.GetSymbolInfo(parameter.Type).Symbol as ITypeSymbol }
+                        : new[] { (ITypeSymbol)null };
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: I'd prefer `default(ITypeSymbol)`, but it's not worth to change it now - more a comment for future.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/AvoidExcessiveClassCoupling.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+{
+    interface IFoo { }
+    class FooBase : IFoo { }
+    class Foo1 : FooBase { }
+    public struct MyStruct { }
+
+    public abstract class TestCases // Noncompliant {{Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 7 to the maximum authorized 1 or less.}}
+//                        ^^^^^^^^^
+    {
+        // ================================================================================
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/AvoidExcessiveClassCoupling.cs]: That's a bit funny formatting :)

Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/AvoidExcessiveClassCoupling.cs]: It was a pain to read the different parts when I was doing my tests. I could actually remove the sections now if you guys want.

Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/AvoidExcessiveClassCoupling.cs]: Fine with me to leave it.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs (hunk 4 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: Cannot update a sub-task on RSPEC.

Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: Then flag it to PMs and let them handle it.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs (hunk 2 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: I made a test for fields and `Task`, `Action` and `Func` do count. I will update my tests to see if I correctly handle them (should be good). Any idea of other special types? I will make a test with pointers.

Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: Nope.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+            {
+                yield return model.GetSymbolInfo(identifierName).Symbol as ITypeSymbol;
+            }
+        }
+
+        private static int FilterAndCountCoupling(IEnumerable<ITypeSymbol> types)
+        {
+            return types.Distinct()
+                .WhereNotNull()
+                .Where(type => type.TypeKind != TypeKind.Enum &&
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: MAJOR Drop 'Where' and move the condition into the 'Count'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2971)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs (hunk 7 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs]: MAJOR Drop 'Where' and move the condition into the 'Count'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2971)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**